Add an inplace=False/True keyword argument to ``__call__`` on fitters to allow in-place modification of models

astrofrog opened the pull request:
This then uses this keyword argument to speed up the fitting in ``parallel_fit_dask`` since we can modify the parameters in-place without any risk.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

pllim commented on docs/changes/modeling/17033.feature.rst:
Should this be in "perf" too?

astrofrog replied:
Yes I think so

astrofrog replied:
Done

Cadair commented:
pre-commit.ci autofix

mhvk reviewed (commented):
This seems like the nicer approach! I guess one small suggestion (but might as well wait for proper review), is to add a `*` to ensure the new argument is keyword-only.

larrybradley commented:
I'm not a `modeling` maintainer, but I like this addition!

astrofrog commented:
> This seems like the nicer approach! I guess one small suggestion (but might as well wait for proper review), is to add a `*` to ensure the new argument is keyword-only.

Done!

nden commented:
This looks good. A couple of comments
- Can you edit the `Returns` sections in the docstrings to clarify that the returned model is not always a copy
- Should the `inplace` parameter be passed to the `FittingWithOutlierRemoval`? Speeding up this one will be a good thing too.

nden commented:
Also I wonder if a note should be added to the `Performance tips` section in the documentation.
https://docs.astropy.org/en/latest/modeling/performance.html

astrofrog commented:
@nden - I've now:

* Added a performance tip
* Updated the Returns sections in the docstrings for all the fitters
* Added an explicit inplace= option to ``FittingWithOutlierRemoval`` (in fact the option was passed through anyway via ``**kwargs`` but I think it's clearer this way

astrofrog commented:
pre-commit.ci autofix

astrofrog commented:
@nden - this should be ready, the CI failures are unrelated.

astrofrog commented:
Rebased
